controllers/helpers.go

```diff
@@ -121,3 +129,215 @@ func GetObjectsToRequestsByNamespaceAndClusterName(ctx context.Context, c client
+func referSameObject(a, b metav1.OwnerReference) bool {
+	aGV, err := schema.ParseGroupVersion(a.APIVersion)
+	if err != nil {
+		return false
+	}
+
+	bGV, err := schema.ParseGroupVersion(b.APIVersion)
+	if err != nil {
+		return false
+	}
+
+	return aGV.Group == bGV.Group && a.Kind == b.Kind && a.Name == b.Name
+}
+
+// GetCloudProviderSecret returns the required azure json secret for the provided parameters. It differentiates between
+func GetCloudProviderSecret(d azure.ClusterDescriber, namespace, name string, identityType infrav1.VMIdentity, userIdentityID string) (*corev1.Secret, error) {
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      fmt.Sprintf("%s-azure-json", name),
+			Labels: map[string]string{
+				string(infrav1.ResourceLifecycleOwned): "",
+			},
+		},
+	}
+
+	var data []byte
+	var err error
+
+	switch identityType {
+	case infrav1.VMIdentitySystemAssigned:
+		data, err = systemAssignedIdentityCloudProviderConfig(d)
+		if err != nil {
+			return nil, err
+		}
+	case infrav1.VMIdentityUserAssigned:
+		if len(userIdentityID) < 1 {
+			return nil, errors.New("expected a non-empty userIdentityID")
+		}
+		data, err = userAssignedIdentityCloudProviderConfig(d, userIdentityID)
+		if err != nil {
+			return nil, err
+		}
+	case infrav1.VMIdentityNone:
+		data, err = servicePrincipalCloudProviderConfig(d)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	secret.Data = map[string][]byte{
+		"azure.json": data,
+	}
+
+	return secret, nil
+}
+
+func servicePrincipalCloudProviderConfig(d azure.ClusterDescriber) ([]byte, error) {
+	return json.MarshalIndent(newCloudProviderConfig(d), "", "    ")
+}
+
+func systemAssignedIdentityCloudProviderConfig(d azure.ClusterDescriber) ([]byte, error) {
+	config := newCloudProviderConfig(d)
+	config.AadClientID = ""
+	config.AadClientSecret = ""
+	config.UseManagedIdentityExtension = true
+	return json.MarshalIndent(config, "", "    ")
+}
+
+func userAssignedIdentityCloudProviderConfig(d azure.ClusterDescriber, identityID string) ([]byte, error) {
+	config := newCloudProviderConfig(d)
+	config.AadClientID = ""
+	config.AadClientSecret = ""
+	config.UseManagedIdentityExtension = true
+	config.UserAssignedIdentityID = identityID
+	return json.MarshalIndent(config, "", "    ")
+}
+
+func newCloudProviderConfig(d azure.ClusterDescriber) *CloudProviderConfig {
+	return &CloudProviderConfig{
+		Cloud:                        d.CloudEnvironment(),
+		AadClientID:                  d.ClientID(),
+		AadClientSecret:              d.ClientSecret(),
+		TenantID:                     d.TenantID(),
+		SubscriptionID:               d.SubscriptionID(),
+		ResourceGroup:                d.ResourceGroup(),
+		SecurityGroupName:            d.NodeSubnet().SecurityGroup.Name,
+		Location:                     d.Location(),
+		VMType:                       "vmss",
+		VnetName:                     d.Vnet().Name,
+		VnetResourceGroup:            d.Vnet().ResourceGroup,
```

Comment:
Can you add the NSG group in there as well to fix https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/356?

Reply:
there's no way for users to override right now, correct? we're just pulling from azure cluster

Reply:
There is, if you bring your own vnet and the NSG is in the vnet rg which is different from the cluster rg

Reply:
I don't think that's implied, I think all three can be different? I just went to the Azure portal and picked a subnet from a vnet in one RG, it allows me to associate with an NSG from other resource groups.

If I change this here, even if vnet + nsg is the more common scenario, we're trading the existing bug for a new one. we probably need to add nsg RG to the spec somewhere?

Reply:
either way happy to expose the value and set it to d.ResourceGroup(), then we can follow up and allow setting it to other things?

Reply:
> If I change this here, even if vnet + nsg is the more common scenario, we're trading the existing bug for a new one. we probably need to add nsg RG to the spec somewhere?

💯 

let's hold then